Add txid + nout

kaykurokawa opened the pull request:
This is the rebased pull request for adding txid + nout into lbrynet.  

Requires lbryum pull request : 
https://github.com/lbryio/lbryum/pull/34

jobevers commented on lbrynet/core/Wallet.py:
```diff
@@ -313,12 +314,12 @@ def get_txid_for_name(self, name):
         def get_claim_for_name(claims):
             for claim in claims:
-                if claim['name'] == name:
+                if claim['name'] == name and claim['nOut'] == nout:
```
this logic gets repeated, and I generally dislike complicated if statements.  This might be better if there was a data structure, maybe:

```
ClaimLocation = collections.namedtuple('ClaimIdentifier', 'txid nout')
```

And then changed each function to take in a `claim_location` like

```
def get_stream_info_from_txid(self, name, claim_location):
```

and have a function:

```
def matchesClaim(claim, claim_location):
    return claim['name'] == claim_location.name and claim['nOut'] == claim_location.nout
```

kaykurokawa replied:
This is a good idea, but I'm gonna hold this off as the last thing to do, since its going to change the interface to a bunch of functions and will take some time to code. If I don't have time for it this week, it will have to go into next release cycle.

jobevers commented on lbrynet/core/Wallet.py:
```diff
@@ -313,12 +314,12 @@ def get_txid_for_name(self, name):
         return d
 
-    def get_stream_info_from_txid(self, name, txid):
+    def get_stream_info_from_txid(self, name, txid, nout):
```
might be worth changing the name of this - `get_stream_info_from_claim` maybe?

kaykurokawa replied:
Changed to get_stream_info_from_claim_outpoint as outpoint is the term used in bitcoin to refer to txid + nout

jobevers commented on lbrynet/core/Wallet.py:
```diff
@@ -529,35 +511,16 @@ def address_is_mine(self, address):
     def get_name_and_validity_for_sd_hash(self, sd_hash):
         d = self._get_claim_metadata_for_sd_hash(sd_hash)
-        d.addCallback(lambda name_txid: self._get_status_of_claim(name_txid[1], name_txid[0], sd_hash) if name_txid is not None else None)
+        d.addCallback(lambda name_txid: self._get_status_of_claim(name_txid[1], name_txid[2], name_txid[0], sd_hash) if name_txid is not None else None)
```
I think this is worth making a small nested function to increase readability:

```
def _get_status_of_claim(name_txid, sd_hash):
    if name_txid:
        return self._get_status_of_claim(self._get_status_of_claim(name_txid[1], name_txid[2], name_txid[0], sd_hash))
    else:
        return None
```

kaykurokawa replied:
Changed

jobevers commented on lbrynet/db_migrator/dbmigrator.py:
```diff
@@ -1,18 +1,14 @@
+            from lbrynet.db_migrator.migrate1to2 import do_migration
+            do_migration(db_dir)
+            current += 1
+    return None
```
I don't understand this logic.  What if current = 2 and end = 3, this will loop forever.

jobevers commented on lbrynet/db_migrator/migrate1to2.py:
```diff
@@ -0,0 +1,72 @@
+                       "    claimId text, " +
+                       "    name text, " +
+                       "    txid text, " +
+                       "    n integer)")
```
the trailing `+` isn't necessary.  For example:

```
my_string = ('asldkjfasldkfjasdlfkjasdf' 
                      'alskdjflaskjdfa')
```

is valid.

kaykurokawa replied:
Changed

jobevers commented on lbrynet/lbryfilemanager/EncryptedFileDownloader.py:
```diff
@@ -42,35 +42,37 @@ def __init__(self, rowid, stream_hash, peer_finder, rate_limiter, blob_manager,
+            err_msg = "Error loading name claim for lbry file: \
+                       lbry://%s, tx %s output %i does not contain a valid claim"
+            log.error(err_msg, name, txid, nout)
+            log.warning("lbry file for lbry://%s, tx %s output %i has no claim, deleting it", name, txid, nout)
```
the warning and error seem redundant.  Remove the warning statement

kaykurokawa replied:
Changed

jobevers commented on lbrynet/lbrynet_daemon/Daemon.py:
```diff
@@ -474,10 +474,13 @@ def _check_connection_problems(self):
-        self.pending_claims[name] = txid
-        return txid
+    # claim_out is dictionary containing 'txid' and 'nout' 
+    def _add_to_pending_claims(self, name, claim_out):
```
can replace `claim_out` with the `ClaimLocation` data structure

jobevers commented on lbrynet/lbrynet_daemon/Daemon.py:
```diff
@@ -711,29 +717,30 @@ def _setup_data_directory(self):
 
+        if not os.path.exists(os.path.join(self.db_dir, "db_revision")):
+            log.warning("db_revision file not found")
+            db_revision = open(os.path.join(self.db_dir, "db_revision"), mode='w')
```
use `with`:  
```
with open(...) as db_revision: 
    db_revision.write(str(old_revision))
```

kaykurokawa replied:
Changed, along with adding db_revision file to configuration file

jobevers commented on lbrynet/lbrynet_daemon/Daemon.py:
```diff
@@ -711,29 +717,30 @@ def _setup_data_directory(self):
+
+        def update_version_file_and_print_success():
+            with open(os.path.join(self.db_dir, "db_revision"), mode='w') as db_revision:
+                db_revision.write(str(self.current_db_revision))
```
Don't Repeat Yourself

kaykurokawa replied:
Changed

jobevers commented on lbrynet/lbrynet_daemon/Daemon.py:
```diff
@@ -2538,27 +2563,73 @@ def wallet(self):
+                return True
+            return False
+        lbry_file = next(lf for lf in self.daemon.lbry_file_manager.lbry_files if lf.uri == self.name)
+        d = _remove_if_should(lbry_file)
```
I whole-heartedly dislike this pattern; nested functions are gross. I much more would like to see this functionality moved into a separate class.

```
class Remover(object):
    def __init__(self, daemon, lbry_file):
        self.daemon = daemon
        self.lbry_file

    def remove_if_should(self):
        ...

     def get_and_check_tx(self, txid):
       ...
```

and then this function looks like:

```
def _confirm_should_remove(self):
    lbry_file = next(lf for lf in self.daemon.lbry_file_manager.lbry_files if lf.uri == self.name)
    d =  Remover(self, lbry_file).remove_if_should()
    return d
```

kaykurokawa replied:
I think we need to revert ResolveNameHelper class back to what it was before, after reviewing I don't think the changes are relevant to this patch.

jobevers commented on lbrynet/lbrynet_daemon/Daemon.py:
```diff
@@ -2538,27 +2563,73 @@ def wallet(self):
+        d = self.wallet.get_name_claims()
+        d.addCallback(lambda claims: self.get_my_valid_claim(claims, self.name))
+        d.addCallback(lambda r: self._add_txid() if not r else r)
+        d.addCallbacks(lambda r: self.wallet.get_stream_info_for_name(self.name) if r else UnknownNameError)
```
I think UnknownNameError needs to be wrapped in a defer.Fail?

kaykurokawa replied:
Changed

jobevers commented on lbrynet/lbrynet_daemon/Publisher.py:
```diff
@@ -38,8 +39,11 @@ def __init__(self, session, lbry_file_manager, wallet):
         def _show_result():
-            log.info("Success! Published %s --> lbry://%s txid: %s", self.file_name, self.publish_name, self.txid)
-            return defer.succeed(self.txid)
+            log.info("Success! Published %s --> lbry://%s txid: %s nout: %d", self.file_name, self.publish_name, self.txid, self.nout)
```
nit: wrap the long line (I'll add a pylint check for this at some point... I just don't feel like cleaning up all of them...)

kaykurokawa replied:
Changed

jobevers commented:
There is a lot here!  Hopefully I didn't lose too much focus near the end.

We don't have a good test framework yet (its improving) but some tests, especially in the db migration would be comforting.  I'll leave it up to you to decide if there is other high risk parts of the change that warrant testing.

kaykurokawa commented:
Made changes based on job's review. One thing I did not do is make a new data structure to represent txid + nout as that is going to take a bit of time ( will try to do either by end of this week or for next release cycle). 

Found some bugs, and fixed it. 

Reverted jack's changes to _ResolveNameHelper as it does not seem relevant to this patch.

kaykurokawa commented:
Rebased to master

kaykurokawa commented:
https://github.com/lbryio/lbry/pull/256 should be merged to master first , and this pull request rebased

jobevers commented:
regarding a ClaimLocation/Outpoint data structure - do you want to make this part of this PR or save it for future work?

kaykurokawa commented:
Since this didn't make it into this week's release cycle, I can make ClaimLocation/Outpoint data structure part of this PR

lyoshenka commented:
@kaykurokawa when can this get merged in. its been open for a long time. whats the next step?

kaykurokawa commented:
I finished the ClaimOutpoint data structure. I'm rebasing to master now and testing, will be ready to review by end of today.

kaykurokawa commented:
Finished adding the ClaimOutpoint data structure, which is used internally in the Wallet class as a way to encapsulate txid and nout.

Needs final review for merging.

jobevers commented:
lgtm

kaykurokawa commented:
Rebased to master and squashed to a single commit. 
Merging to master.
